crates/metadata/src/compatibility.rs

```diff
@@ -0,0 +1,177 @@
+// Copyright 2018-2023 Parity Technologies (UK) Ltd.
+// This file is part of cargo-contract.
+//
+// cargo-contract is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// cargo-contract is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with cargo-contract.  If not, see <http://www.gnu.org/licenses/>.
+
+use anyhow::{
+    anyhow,
+    bail,
+    Result,
+};
+use semver::{
+    BuildMetadata,
+    Comparator,
+    Op,
+    Version,
+    VersionReq,
+};
+
+/// Version of the currently executing `cargo-contract` binary.
+const VERSION: &str = env!("CARGO_PKG_VERSION");
+
+#[derive(Debug, serde::Deserialize)]
+struct CompatibilityList {
+    versions: Vec<CompatibilityInfo>,
+}
+
+#[derive(Debug, serde::Deserialize)]
+struct CompatibilityInfo {
+    #[serde(rename = "cargo-contract")]
+    cargo_contract_req: VersionReq,
+    #[serde(rename = "ink")]
+    ink_req: VersionReq,
```

Comment:
How about making this a `Vec<VersionReq>` and making it a logical OR? Then we can provide additional requirements for prerelease ranges.

```json
{
    "cargo-contract": "2.0.0",
    "ink": [
        "4.0.0-alpha.3",
        "4.0.0"
    ]
}
```

Reply:
Done